src/fomo.jl

```diff
@@ -0,0 +1,367 @@
+      max_eval = max_eval,
+      iter = stats.iter,
+      max_iter = max_iter,
+      max_time = max_time,
+    ),
+  )
+
+  callback(nlp, solver, stats)
+
+  done = stats.status != :unknown
+
+  d .= ∇fk
+  norm_d = norm_∇fk
+  βmax = T(0)
+  ρk = T(0)
+  avgβmax = T(0)
+  siter = 0
+  oneT = T(1)
+  while !done
+    λk = step_mult(αk,norm_d,backend)
+    c .= x .- λk .* d
+    step_underflow = x == c # step addition underfow on every dimensions, should happen before αk == 0
+    if r2mode
+      ΔTk = norm_∇fk^2 * λk
+    else  
+      ΔTk = dot(∇fk , d) * λk
+    end
+    fck = obj(nlp, c)
+    if fck == -Inf
+      set_status!(stats, :unbounded)
+      break
+    end
+    ρk = (stats.objective - fck) / ΔTk
+    # Update regularization parameters
+    if ρk >= η2
+      αk = min(αmax, γ2 * αk)
+    elseif ρk < η1
+      αk = αk * γ1
+      if !r2mode
+        βmax *= γ3
+        d .= ∇fk .* (oneT - βmax) .+ m .* βmax
+      end
+    end
+
+    # Acceptance of the new candidate
+    if ρk >= η1
+      x .= c
+      if !r2mode
+        m .= ∇fk .* (oneT - β) .+ m .* β
+      end
+      set_objective!(stats, fck)
+      grad!(nlp, x, ∇fk)
+      norm_∇fk = norm(∇fk)
+      if !r2mode
+        βmax = find_beta(m, ∇fk, norm_∇fk, β, θ1, θ2)
+        d .= ∇fk .* (oneT - βmax) .+ m .* βmax
+        norm_d = norm(d)
+      else
+        d .= ∇fk
+        norm_d = norm_∇fk
+      end
+      if !r2mode
+        avgβmax += βmax
+        siter += 1
+      end
+    end
+
+    set_iter!(stats, stats.iter + 1)
+    set_time!(stats, time() - start_time)
+    set_dual_residual!(stats, norm_∇fk)
+    optimal = norm_∇fk ≤ ϵ
+
+    if verbose > 0 && mod(stats.iter, verbose) == 0
+      @info infoline
+      if r2mode
+        infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk 1/αk ρk
+      else
+        infoline = @sprintf "%5d  %9.2e  %7.1e  %7.1e  %7.1e  %7.1e" stats.iter stats.objective norm_∇fk αk ρk βmax
+      end
+    end
+
+    set_status!(
+      stats,
+      get_status(
+        nlp,
+        elapsed_time = stats.elapsed_time,
+        optimal = optimal,
+        max_eval = max_eval,
+        iter = stats.iter,
+        max_iter = max_iter,
+        max_time = max_time,
+      ),
+    )
+    
+    callback(nlp, solver, stats)
+
+    step_underflow  && set_status!(stats,:small_step)
+    αk == 0         && set_status!(stats,:exception) # :small_nlstep exception should happen before
+
+    done = stats.status != :unknown
+  end
+  if !r2mode
+    avgβmax /= siter
+    stats.solver_specific[:avgβmax] = avgβmax
+  end
+  set_solution!(stats, x)
+  return stats
+end
+
+"""
+find_beta(m, ∇f, norm_∇f, β, θ1, θ2)
+
+Compute βmax which saturates the contibution of the momentum term to the gradient.
+`βmax` is computed such that the two gradient-related conditions are ensured: 
+1. [(1-βmax) * ∇f(xk) + βmax * dot(m,∇f(xk))] ≥ θ1 * ||∇f(xk)||²
+2. ||∇f(xk)|| ≥ θ2 * ||(1-βmax) * ∇f(xk) + βmax * dot(m.∇f(xk))||
```

Comment:
It doesn’t look like you’re using the “norm” symbol (type `\Vert` and then TAB).